framework/src/dgkernels/DGLowerDKernel.C

```diff
@@ -0,0 +1,254 @@
+
+      // Compute neighbor-neighbor Jacobian
+      computeOffDiagElemNeighJacobian(Moose::NeighborNeighbor, jvar);
+
+      // Compute the other five pieces of Jacobian related with lower-d variable
+      computeOffDiagLowerDJacobian(Moose::LowerLower, jvar);
+      computeOffDiagLowerDJacobian(Moose::LowerSecondary, jvar);
+      computeOffDiagLowerDJacobian(Moose::LowerPrimary, jvar);
+      computeOffDiagLowerDJacobian(Moose::SecondaryLower, jvar);
+      computeOffDiagLowerDJacobian(Moose::PrimaryLower, jvar);
+    }
+  }
+}
+
+void
+DGLowerDKernel::computeOffDiagLowerDJacobian(Moose::ConstraintJacobianType type,
+                                             const MooseVariableFEBase & jvar)
+{
+  mooseAssert(type != Moose::PrimaryPrimary && type != Moose::PrimarySecondary &&
+                  type != Moose::SecondaryPrimary && type != Moose::SecondarySecondary,
+              "Jacobian types without lower should be handled in computeOffDiagElemNeighJacobian");
+
+  const VariableTestValue & test_space =
+      (type == Moose::LowerLower || type == Moose::LowerSecondary || type == Moose::LowerPrimary)
+          ? _test_lambda
+          : (type == Moose::PrimaryLower ? _test : _test_neighbor);
+  unsigned int ivar =
+      (type == Moose::LowerLower || type == Moose::LowerSecondary || type == Moose::LowerPrimary)
+          ? _lowerd_var.number()
+          : _var.number();
+
+  prepareMatrixTagLower(_assembly, ivar, jvar.number(), type);
+  if (_local_ke.n() == 0 || _local_ke.m() == 0)
+    return;
+
+  if (jvar.fieldType() == Moose::VarFieldType::VAR_FIELD_STANDARD)
+  {
+    const auto & jv0 = static_cast<const MooseVariable &>(jvar);
+    const VariableTestValue & loc_phi =
+        (type == Moose::LowerLower || type == Moose::SecondaryLower || type == Moose::PrimaryLower)
+            ? jv0.phiLower()
+            : (type == Moose::LowerPrimary ? jv0.phiFace() : jv0.phiFaceNeighbor());
+
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      initLowerDQpOffDiagJacobian(type, jvar);
+      for (_i = 0; _i < test_space.size(); _i++)
+        for (_j = 0; _j < loc_phi.size(); _j++)
+        {
+          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobian(type, jvar);
+        }
+    }
+  }
+  else if (jvar.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)
+  {
+    const auto & jv1 = static_cast<const MooseVariable &>(jvar);
+    const VariableTestValue & loc_phi =
+        (type == Moose::LowerLower || type == Moose::SecondaryLower || type == Moose::PrimaryLower)
+            ? jv1.phiLower()
+            : (type == Moose::LowerPrimary ? jv1.phiFace() : jv1.phiFaceNeighbor());
+
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      initLowerDQpOffDiagJacobian(type, jvar);
+      for (_i = 0; _i < test_space.size(); _i++)
+        for (_j = 0; _j < loc_phi.size(); _j++)
+        {
+          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobian(type, jvar);
```

Comment:
This does not seem right. Possibly you want to check out `Kernel::computeOffDiagJacobian` for handling coupling array variables into a normal kernel.

Reply:
But looks like `DGKernel::computeOffDiagElemNeighJacobian` is not doing things right either. Maybe we can just throw an error in this `if (jvar.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)` branch. @lindsayad do you have suggestions?

Reply:
Basically, I made this part based on `DGKernel::computeOffDiagElemNeighJacobian`. Can you give me more specific guide to correct it?

Reply:
`Kernel::computeOffDiagJacobian` should be the place to check.

Reply:
Know I see. You are expecting to have something like this:
`DGLowerDKernel::computeOffDiagLowerDJacobian`
```C
      for (_i = 0; _i < test_space.size(); _i++)
        for (_j = 0; _j < loc_phi.size(); _j++)
        {
         RealEigenVector v = _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobianArray(type, jvar);
         _assembly.saveFullLocalArrayJacobian(
              _local_ke, _i, test_space.size(), _j, loc_phi.size(), ivar, jvar.number(), v);
        }
```

`HFEMDirichletBC::computeLowerDQpOffDiagJacobianArray`
```C
RealEigenMatrix
HFEMDirichletBC::computeLowerDQpOffDiagJacobianArray(Moose::ConstraintJacobianType type,
                                                     const MooseVariableFEBase & jvar)
{
  if (_uhat_var && jvar.number() == _uhat_var->number() && type == Moose::LowerLower)
    return -_phi[_j][_qp] * _test_lambda[_i][_qp];
  else
    return RealEigenMatrix::Zero(_count, jvar.count());
}
```

Reply:
Yes. But we may not add this and just throw an error. We can add it later if needs come in the future.

Reply:
I updated to just throw an error.